### PR TITLE
Iterator range maintenance

### PR DESCRIPTION
**Pull request by paplorinc**
Fixes https://github.com/javaslang/javaslang/issues/1309

### CODE REVIEW

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 1]
```diff
@@ -467,52 +467,16 @@ public Short getNext() {
...
-                        curr += step;
-                        if (curr == prev) {
-                            hasNext = false;
-                        }
-                    }
-                    return next;
-                }
-            };
-        }
+        final double toInclusive = (step > 0) ? Math.nextDown(toExclusive) : Math.nextUp(toExclusive);
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: make it inclusive by incrementing it with a `ULP`

Note: this will make exclusive infinities not possible (had to modify a test for that), but I think an exclusive equality to infinity is strange anyway

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 2]
```diff
@@ -658,36 +622,45 @@ public Double getNext() {
...
+
+                    @Override
+                    public boolean hasNext() {
+                        return i > end;
+                    }
 
-                @Override
-                public Double getNext() {
-                    final double next = curr;
-                    if ((step > 0 && curr + step > toInclusive) || (step < 0 && curr + step < toInclusive)) {
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Instead of checking for constant conditions inside the loop, let's make two versions, and ascending and a descending -- simplified implementation significantly

File: javaslang/src/test/java/javaslang/collection/AbstractTraversableRangeTest.java [hunk 3]
```diff
@@ -547,29 +547,4 @@ public void shouldProhibitDoubleRangeByFromEqualNaN() {
...
-
-    @Test
-    public void shouldCreateDoubleRangeByFromMaxToInfinity() {
-        assertThat(rangeBy(Double.MAX_VALUE, Double.POSITIVE_INFINITY, 3E307)).isEqualTo(of(Double.MAX_VALUE));
-        assertThat(rangeBy(-Double.MAX_VALUE, Double.NEGATIVE_INFINITY, -3E307)).isEqualTo(of(-Double.MAX_VALUE));
-    }
-
-    @Test
-    public void shouldCreateDoubleRangeClosedByFromMaxToInfinity() {
-        assertThat(rangeClosedBy(Double.MAX_VALUE, Double.POSITIVE_INFINITY, 3E307)).isEqualTo(of(Double.MAX_VALUE, Double.POSITIVE_INFINITY));
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/AbstractTraversableRangeTest.java]: these tests have slightly changed, please compare them against the new impl

**Comment by danieldietrich** [comment]
As described in #1309 these methods should work exactly as in Scala, especially the corner cases (Infinity, NaN, ...).

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 4]
```diff
@@ -467,51 +470,79 @@ public Short getNext() {
...
                     }
-                    return next;
-                }
-            };
+                };
+            }
+        }
+    }
+
+    // TODO these two need a place to crash for the night :/
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: @danieldietrich, these guys need to sleep somewhere

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 5]
```diff
@@ -467,51 +470,79 @@ public Short getNext() {
...
-        } else if (Double.isNaN(toExclusive)) {
-            throw new IllegalArgumentException("toExclusive is NaN");
-        } else if (Double.isNaN(step)) {
-            throw new IllegalArgumentException("step is NaN");
-        } else if (step == 0) {
+        final BigDecimal fromDecimal = asDecimal(from), toDecimal = asDecimal(toExclusive), stepDecimal = asDecimal(step);
+        return rangeBy(fromDecimal, toDecimal, stepDecimal);
+    }
+
+    static Iterator<Double> rangeBy(BigDecimal from, BigDecimal toExclusive, BigDecimal step) {
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Here I would expect an `Iterator<BigDecimal>` as result.

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 6]
```diff
@@ -467,51 +470,79 @@ public Short getNext() {
...
     static Iterator<Double> rangeBy(double from, double toExclusive, double step) {
-        if (Double.isNaN(from)) {
-            throw new IllegalArgumentException("from is NaN");
-        } else if (Double.isNaN(toExclusive)) {
-            throw new IllegalArgumentException("toExclusive is NaN");
-        } else if (Double.isNaN(step)) {
-            throw new IllegalArgumentException("step is NaN");
-        } else if (step == 0) {
+        final BigDecimal fromDecimal = asDecimal(from), toDecimal = asDecimal(toExclusive), stepDecimal = asDecimal(step);
+        return rangeBy(fromDecimal, toDecimal, stepDecimal);
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: could return `rangeBy(fromDecimal, toDecimal, stepDecimal).map(BigDecimal::doubleValue)`

File: javaslang/src/main/java/javaslang/collection/Iterator.java (hunk 5 above)
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: I guess my helper methods aren't the only one who don't know where to sleep :p

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 7]
```diff
@@ -467,51 +470,79 @@ public Short getNext() {
...
+
+    // TODO these two need a place to crash for the night :/
+
+    /* scale-independent equality */
+    static boolean areEqual(BigDecimal from, BigDecimal toExclusive) {
+        return from.compareTo(toExclusive) == 0;
+    }
+
+    /* parse infinite values also */
+    static final BigDecimal infinityMultiplier = BigDecimal.valueOf(1.000000000000001); // decrease by as little as possible so that it doesn't have a double representation anymore -> infinite
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: instead of multiplying we could use `movePointRight(1)` - then this const can be removed

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 8]
```diff
@@ -467,51 +470,79 @@ public Short getNext() {
...
+
+    static BigDecimal asDecimal(double number) {
+        if (number == NEGATIVE_INFINITY) {
+            final BigDecimal result = BigDecimal.valueOf(Math.nextUp(NEGATIVE_INFINITY));
+            return result.multiply(infinityMultiplier);
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: return result.movePointRight(1);

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 9]
```diff
@@ -467,51 +470,79 @@ public Short getNext() {
...
+        } else if (number == POSITIVE_INFINITY) {
+            final BigDecimal result = BigDecimal.valueOf(Math.nextDown(POSITIVE_INFINITY));
+            return result.multiply(infinityMultiplier);
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: dito

File: javaslang/src/main/java/javaslang/collection/Iterator.java (hunk 8 above)
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: That probably won't be the smallest value which has infinite for `toDouble`, but it might not really matter anyway :)

Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Yeah, may tests are failing that way, e.g. `shouldCreateRangeClosedByWhereFromIsLessThanToAndStepCorrectDirection`

File: javaslang/src/main/java/javaslang/collection/Iterator.java (hunk 5 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: yes, no sleep here. topics are too interesting :)

File: javaslang/src/main/java/javaslang/collection/Iterator.java (hunk 8 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Mmhh, I've looked at the tests - don't see a reason why they might fail. Do you know which one of them fails?

```java
    @Test
    public void shouldCreateRangeClosedByWhereFromIsLessThanToAndStepCorrectDirection() {

        // char
        assertThat(rangeClosedBy('a', 'c', 1)).isEqualTo(of('a', 'b', 'c'));
        assertThat(rangeClosedBy('a', 'e', 2)).isEqualTo(of('a', 'c', 'e'));
        assertThat(rangeClosedBy('a', 'f', 2)).isEqualTo(of('a', 'c', 'e'));
        assertThat(rangeClosedBy((char) (Character.MAX_VALUE - 2), Character.MAX_VALUE, 3)).isEqualTo(of((char) (Character.MAX_VALUE - 2)));
        assertThat(rangeClosedBy((char) (Character.MAX_VALUE - 3), Character.MAX_VALUE, 3)).isEqualTo(of((char) (Character.MAX_VALUE - 3), Character.MAX_VALUE));
        assertThat(rangeClosedBy('c', 'a', -1)).isEqualTo(of('c', 'b', 'a'));
        assertThat(rangeClosedBy('e', 'a', -2)).isEqualTo(of('e', 'c', 'a'));
        assertThat(rangeClosedBy('e', (char) ('a' - 1), -2)).isEqualTo(of('e', 'c', 'a'));
        assertThat(rangeClosedBy((char) (Character.MIN_VALUE + 2), Character.MIN_VALUE, -3)).isEqualTo(of((char) (Character.MIN_VALUE + 2)));
        assertThat(rangeClosedBy((char) (Character.MIN_VALUE + 3), Character.MIN_VALUE, -3)).isEqualTo(of((char) (Character.MIN_VALUE + 3), Character.MIN_VALUE));

        // double
        assertThat(rangeClosedBy(1.0, 3.0, 1.0)).isEqualTo(of(1.0, 2.0, 3.0));
        assertThat(rangeClosedBy(1.0, 5.0, 2.0)).isEqualTo(of(1.0, 3.0, 5.0));
        assertThat(rangeClosedBy(1.0, 6.0, 2.0)).isEqualTo(of(1.0, 3.0, 5.0));
        assertThat(rangeClosedBy(Double.MAX_VALUE - 2.0E307, Double.MAX_VALUE, 3.0E307)).isEqualTo(of(Double.MAX_VALUE - 2.0E307));
        assertThat(rangeClosedBy(3.0, 1.0, -1.0)).isEqualTo(of(3.0, 2.0, 1.0));
        assertThat(rangeClosedBy(5.0, 1.0, -2.0)).isEqualTo(of(5.0, 3.0, 1.0));
        assertThat(rangeClosedBy(5.0, 0.0, -2.0)).isEqualTo(of(5.0, 3.0, 1.0));
        assertThat(rangeClosedBy(-Double.MAX_VALUE + 2.0E307, -Double.MAX_VALUE, -3.0E307)).isEqualTo(of(-Double.MAX_VALUE + 2.0E307));

        // int
        assertThat(rangeClosedBy(1, 3, 1)).isEqualTo(of(1, 2, 3));
        assertThat(rangeClosedBy(1, 5, 2)).isEqualTo(of(1, 3, 5));
        assertThat(rangeClosedBy(1, 6, 2)).isEqualTo(of(1, 3, 5));
        assertThat(rangeClosedBy(Integer.MAX_VALUE - 2, Integer.MAX_VALUE, 3)).isEqualTo(of(Integer.MAX_VALUE - 2));
        assertThat(rangeClosedBy(Integer.MAX_VALUE - 3, Integer.MAX_VALUE, 3)).isEqualTo(of(Integer.MAX_VALUE - 3, Integer.MAX_VALUE));
        assertThat(rangeClosedBy(3, 1, -1)).isEqualTo(of(3, 2, 1));
        assertThat(rangeClosedBy(5, 1, -2)).isEqualTo(of(5, 3, 1));
        assertThat(rangeClosedBy(5, 0, -2)).isEqualTo(of(5, 3, 1));
        assertThat(rangeClosedBy(Integer.MIN_VALUE + 2, Integer.MIN_VALUE, -3)).isEqualTo(of(Integer.MIN_VALUE + 2));
        assertThat(rangeClosedBy(Integer.MIN_VALUE + 3, Integer.MIN_VALUE, -3)).isEqualTo(of(Integer.MIN_VALUE + 3, Integer.MIN_VALUE));

        // long
        assertThat(rangeClosedBy(1L, 3L, 1)).isEqualTo(of(1L, 2L, 3L));
        assertThat(rangeClosedBy(1L, 5L, 2)).isEqualTo(of(1L, 3L, 5L));
        assertThat(rangeClosedBy(1L, 6L, 2)).isEqualTo(of(1L, 3L, 5L));
        assertThat(rangeClosedBy(Long.MAX_VALUE - 2, Long.MAX_VALUE, 3)).isEqualTo(of(Long.MAX_VALUE - 2));
        assertThat(rangeClosedBy(Long.MAX_VALUE - 3, Long.MAX_VALUE, 3)).isEqualTo(of(Long.MAX_VALUE - 3, Long.MAX_VALUE));
        assertThat(rangeClosedBy(3L, 1L, -1)).isEqualTo(of(3L, 2L, 1L));
        assertThat(rangeClosedBy(5L, 1L, -2)).isEqualTo(of(5L, 3L, 1L));
        assertThat(rangeClosedBy(5L, 0L, -2)).isEqualTo(of(5L, 3L, 1L));
        assertThat(rangeClosedBy(Long.MIN_VALUE + 2, Long.MIN_VALUE, -3)).isEqualTo(of(Long.MIN_VALUE + 2));
        assertThat(rangeClosedBy(Long.MIN_VALUE + 3, Long.MIN_VALUE, -3)).isEqualTo(of(Long.MIN_VALUE + 3, Long.MIN_VALUE));
    }
```

Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: If we don't provide the minimum value for which double conversion yields infinity, ranges can contain multiple `MAX_VALUE`s, until they reach infinity (and beyond)
infinity and beyond 1

**Comment by danieldietrich** [comment]
Interesting observation regarding that Scala's double range is backed by Double (is it?):

```Scala
  val r = 0d to 4d by 0.1                         //> r  : scala.collection.immutable.NumericRange[Double] = NumericRange(0.0, 0.1,
                                                  //|  0.2, 0.30000000000000004, 0.4, 0.5, 0.6, 0.7, 0.7999999999999999, 0.89999999
                                                  //| 99999999, 0.9999999999999999, 1.0999999999999999, 1.2, 1.3, 1.400000000000000
                                                  //| 1, 1.5000000000000002, 1.6000000000000003, 1.7000000000000004, 1.800000000000
                                                  //| 0005, 1.9000000000000006, 2.0000000000000004, 2.1000000000000005, 2.200000000
                                                  //| 0000006, 2.3000000000000007, 2.400000000000001, 2.500000000000001, 2.60000000
                                                  //| 0000001, 2.700000000000001, 2.800000000000001, 2.9000000000000012, 3.00000000
                                                  //| 00000013, 3.1000000000000014, 3.2000000000000015, 3.3000000000000016, 3.40000
                                                  //| 00000000017, 3.5000000000000018, 3.600000000000002, 3.700000000000002, 3.8000
                                                  //| 00000000002, 3.900000000000002, 4.000000000000002)

	import java.math.BigDecimal
                                                  
  var list: List[Any] = List()                    //> list  : List[Any] = List()
  var b = BigDecimal.valueOf(0d)                  //> b  : java.math.BigDecimal = 0.0
	val step = BigDecimal.valueOf(0.1d)       //> step  : java.math.BigDecimal = 0.1
  while (b.doubleValue() <= 4d) {
      list = list :+ b.doubleValue()
      b = b.add(step)
  }
  println(list)                                   //> List(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1
                                                  //| .4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.
                                                  //| 9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0)
```

File: javaslang/src/main/java/javaslang/collection/Iterator.java (hunk 8 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: It is hard to find the right infinityMultiplyer: `BigDecimal.valueOf(1.0000000000000001110223024625...)`

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Same for adding instead of multiplying:
`BigDecimal.valueOf(Math.nextDown(POSITIVE_INFINITY)).add(new BigDecimal("1.0793...E+292"))`

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Some tests trying to compute the minimal step towards positive infinity:

```java
// = 1.9958403095347198E292
MAX_VALUE - Math.nextDown(MAX_VALUE)

// = 2E+292
BigDecimal.valueOf(MAX_VALUE).subtract(BigDecimal.valueOf(Math.nextDown(MAX_VALUE)))
```

Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Maybe the cleanest way is to extend BigDecimal with representations for positive and negative infinity (e.g. by having an infinity flag)?

Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: The boundaries can quickly be calculated using binary search:

```java
// -- doubles

private BigDecimal binarySearch(BigDecimal finite, BigDecimal infinite) {
    while (true) {
        final BigDecimal magicValue = finite.add(infinite).divide(BigDecimal.valueOf(2), HALF_UP);
        if (Double.isInfinite(magicValue.doubleValue())) {
            if (magicValue.equals(infinite)) {
                return magicValue;
            }
            infinite = magicValue;
        } else {
            finite = magicValue;
        }
    }
}

@Test
public void infinityAndBeyond() {
    final BigDecimal finite = BigDecimal.valueOf(Double.MAX_VALUE);
    final BigDecimal infinite = BigDecimal.valueOf(2).pow(Double.MAX_EXPONENT + 1);

    final BigDecimal negative = binarySearch(finite.negate(), infinite.negate());
    assertThat(negative.add(BigDecimal.ONE).doubleValue()).isNotEqualTo(Double.NEGATIVE_INFINITY);
    assertThat(negative.doubleValue()).isEqualTo(Double.NEGATIVE_INFINITY);

    final BigDecimal positive = binarySearch(finite, infinite);
    assertThat(positive.subtract(BigDecimal.ONE).doubleValue()).isNotEqualTo(Double.POSITIVE_INFINITY);
    assertThat(positive.doubleValue()).isEqualTo(Double.POSITIVE_INFINITY);
}
```

we could calculate and store this value in a static block / field.
I'm still trying to figure out why `179769313486231580793728971405303415079934132710037826936173778980444968292764750946649017977587207096330286416692887910946555547851940402630657488671505820681908902000708383676273854845817711531764475730270069855571366959622842914819860834936475292719074168444365510704342711559699508093042880177904174497792` is so special :p

**Comment by paplorinc** [comment]
> Scala's double range is backed by Double
Quoting from `Range.scala`:
```
  // Double works by using a BigDecimal under the hood for precise
  // stepping, but mapping the sequence values back to doubles with
  // .doubleValue.  This constructs the BigDecimals by way of the
  // String constructor (valueOf) instead of the Double one, which
  // is necessary to keep 0.3d at 0.3 as opposed to
  // 0.299999999999999988897769753748434595763683319091796875 or so.
```

File: javaslang/src/main/java/javaslang/collection/Iterator.java (hunk 8 above)
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: pushed this, @danieldietrich, please advise on where this behavior should be moved

File: javaslang/src/main/java/javaslang/collection/Iterator.java [hunk 10]
```diff
@@ -38,6 +42,42 @@
  * @since 2.0.0
  */
 public interface Iterator<T> extends java.util.Iterator<T>, Traversable<T> {
+    /* TODO these don't belong here! */
```
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Let's move the BigDecimal helper methods down into the `IteratorModule` like this (we need to add some static imports for `areEqual` and `asDecimal`). Please take a look at my TODO comment below.

```java
    final class BigDecimalHelper {

        private static Lazy<BigDecimal> infinityDistance = Lazy.of(() -> {
            final BigDecimal two = BigDecimal.valueOf(2);
            final BigDecimal supremum = BigDecimal.valueOf(Math.nextDown(Double.POSITIVE_INFINITY));
            BigDecimal lowerBound = supremum;
            BigDecimal upperBound = two.pow(Double.MAX_EXPONENT + 1);
            while (true) {
                final BigDecimal magicValue = lowerBound.add(upperBound).divide(two, HALF_UP);
                if (Double.isInfinite(magicValue.doubleValue())) {
                    // TODO: or is areEqual(magicValue, upperBound) more stable?
                    if (magicValue.equals(upperBound)) {
                        return magicValue.subtract(supremum);
                    }
                    upperBound = magicValue;
                } else {
                    lowerBound = magicValue;
                }
            }
        });

        /* scale-independent equality */
        static boolean areEqual(BigDecimal from, BigDecimal toExclusive) {
            return from.compareTo(toExclusive) == 0;
        }

        /* parse infinite values also */
        static BigDecimal asDecimal(double number) {
            if (number == NEGATIVE_INFINITY) {
                final BigDecimal result = BigDecimal.valueOf(Math.nextUp(NEGATIVE_INFINITY));
                return result.subtract(infinityDistance.get());
            } else if (number == POSITIVE_INFINITY) {
                final BigDecimal result = BigDecimal.valueOf(Math.nextDown(POSITIVE_INFINITY));
                return result.add(infinityDistance.get());
            } else {
                return BigDecimal.valueOf(number);
            }
        }
    }
```

For now we do not want to expose these helpers to other classes. Because Java has no module concept (yet), the auxiliary classes (*Module) ensure that they are only accessed from within a class file.

Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Done, thanks.

Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/Iterator.java]: Though I'm not sure how to test it now:
```java
@Test
public void infinityAndBeyond() {
    final BigDecimal negativeInfinity = asDecimal(Double.NEGATIVE_INFINITY);
    assertThat(negativeInfinity.add(BigDecimal.ONE).doubleValue()).isNotEqualTo(Double.NEGATIVE_INFINITY);
    assertThat(negativeInfinity.doubleValue()).isEqualTo(Double.NEGATIVE_INFINITY);

    final BigDecimal positiveInfinity = asDecimal(Double.POSITIVE_INFINITY);
    assertThat(positiveInfinity.subtract(BigDecimal.ONE).doubleValue()).isNotEqualTo(Double.POSITIVE_INFINITY);
    assertThat(positiveInfinity.doubleValue()).isEqualTo(Double.POSITIVE_INFINITY);
}
```

**Comment by danieldietrich** [comment]
I think this PR is also ready now to be pulled in. I will get a last look at the source tomorrow...

**Comment by danieldietrich** [comment]
Thanks!
